File: ios/AtlasCore/Sources/AtlasCore/Spaces/Polls.swift
import Foundation

/// Polls — Sybil-free polling at configurable anonymity levels. Mirrors `backend/atlas/spaces/polls.py`
/// byte-for-byte (Python is reference-of-record). A poll is a structured multi-option vote; one-human-
/// one-response is enforced by a nullifier, and the IdentityTier controls whether a ballot reveals the
/// voter's persona (anonymous = ephemeral ballot key, unlinkable) — never whether one human can stuff it.
extension Spaces {

    fileprivate static let pollDomain = Data("atlas/poll/v1".utf8)
    fileprivate static let pollResponseDomain = Data("atlas/poll-response/v1".utf8)
    fileprivate static let pollIdLabel = Data("atlas/poll-id".utf8)

    fileprivate static func plp(_ d: Data) -> Data {
        var n = UInt32(d.count).bigEndian; var o = Data()
        withUnsafeBytes(of: &n) { o.append(contentsOf: $0) }; o.append(d); return o
    }
    fileprivate static func pu16(_ v: Int) -> Data { var n = UInt16(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }
    fileprivate static func pu32(_ v: Int) -> Data { var n = UInt32(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }
    fileprivate static func pu64(_ v: Int) -> Data { var n = UInt64(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }

    // IdentityTier is the canonical anonymity axis — declared in SpaceKinds.swift, shared with Spaces.

    public struct Poll {
        public var author: HybridSign.PublicKey
        public var question: Data
        public var options: [Data]
        public var tier: IdentityTier
        public var epoch: Int
        public var sig: Data = Data()

        public init(author: HybridSign.PublicKey, question: Data, options: [Data],
                    tier: IdentityTier, epoch: Int, sig: Data = Data()) {
            self.author = author; self.question = question; self.options = options
            self.tier = tier; self.epoch = epoch; self.sig = sig
        }

        func body() -> Data {
            var out = pollDomain + plp(author.encode()) + plp(question) + pu32(options.count)
            for o in options { out += plp(o) }
            out += pu16(tier.rawValue) + pu64(epoch)
            return out
        }
        public func pollID() -> Data { Primitives.H(pollIdLabel, body()) }
    }

    public static func createPoll(_ kp: HybridSign.Keypair, question: Data, options: [Data],
                                  tier: IdentityTier, epoch: Int) throws -> Poll {
        precondition(options.count >= 2, "a poll needs >= 2 options")
        var p = Poll(author: kp.publicKey, question: question, options: options, tier: tier, epoch: epoch)
        p.sig = try HybridSign.sign(kp, p.body())
        return p
    }

    public static func verifyPoll(_ p: Poll) -> Bool {
        p.options.count >= 2 && HybridSign.verify(p.author, p.body(), p.sig)
    }

    public struct PollResponse {
        public var pollID: Data
        public var choice: Int
        public var nullifier: Data
        public var ballotKey: HybridSign.PublicKey   // voter persona OR ephemeral (anonymous)
        public var epoch: Int
        public var sig: Data = Data()

        public init(pollID: Data, choice: Int, nullifier: Data, ballotKey: HybridSign.PublicKey,
                    epoch: Int, sig: Data = Data()) {
            self.pollID = pollID; self.choice = choice; self.nullifier = nullifier
            self.ballotKey = ballotKey; self.epoch = epoch; self.sig = sig
        }

        func body() -> Data {
            pollResponseDomain + plp(pollID) + pu32(choice) + plp(nullifier)
                + plp(ballotKey.encode()) + pu64(epoch)
        }
    }

    static func makeResponse(_ poll: Poll, _ choice: Int, _ nullifier: Data,
                             _ ballotKey: HybridSign.PublicKey, _ epoch: Int) -> PollResponse {
        precondition(choice >= 0 && choice < poll.options.count, "choice out of range")
        return PollResponse(pollID: poll.pollID(), choice: choice, nullifier: nullifier,
                            ballotKey: ballotKey, epoch: epoch)
    }

    /// PSEUDONYMOUS / VERIFIED_PERSON ballot — signed by the voter's persona (choice visible under nym).
    public static func respond(_ voterKp: HybridSign.Keypair, poll: Poll, choice: Int,
                               nullifier: Data, epoch: Int) throws -> PollResponse {
        var r = makeResponse(poll, choice, nullifier, voterKp.publicKey, epoch)
        r.sig = try HybridSign.sign(voterKp, r.body())
        return r
    }

    /// ANONYMOUS ballot — signed by a fresh ephemeral key, unlinkable to the voter's persona; only the
    /// nullifier enforces one-human-one-response. Eligibility is a ps_credential seam in production.
    public static func respondAnonymously(_ poll: Poll, choice: Int, nullifier: Data, epoch: Int,
                                          ephemeralKp: HybridSign.Keypair) throws -> PollResponse {
        var r = makeResponse(poll, choice, nullifier, ephemeralKp.publicKey, epoch)
        r.sig = try HybridSign.sign(ephemeralKp, r.body())
        return r
    }

    public static func verifyResponse(_ poll: Poll, _ r: PollResponse) -> Bool {
        r.pollID == poll.pollID() && r.choice >= 0 && r.choice < poll.options.count
            && HybridSign.verify(r.ballotKey, r.body(), r.sig)
    }

    public struct PollResult {
        public let pollID: Data
        public let counts: [Int]
        public let total: Int
        public func winner() -> Int {
            guard !counts.isEmpty else { return -1 }
            return counts.indices.max(by: { counts[$0] < counts[$1] })!
        }
    }

    /// One-human-one-response: dedupe by nullifier (LAST valid ballot wins). Only valid responses for
    /// THIS poll count.
    public static func tally(_ poll: Poll, _ responses: [PollResponse]) -> PollResult {
        var latest: [Data: PollResponse] = [:]
        for r in responses where verifyResponse(poll, r) {
            latest[r.nullifier] = r
        }
        var counts = [Int](repeating: 0, count: poll.options.count)
        for r in latest.values { counts[r.choice] += 1 }
        return PollResult(pollID: poll.pollID(), counts: counts, total: latest.count)
    }
}

File: backend/atlas/spaces/polls.py
"""Polls over Space content — Sybil-free polling at configurable anonymity levels.

A poll is a structured, multi-option vote. Its integrity rests on the same one-human-one-response
NULLIFIER as social votes (cf. realid/space_pseudonym), so a poll cannot be brigaded by bots or
sockpuppets — one enrolled human, one response, guaranteed. The ANONYMITY LEVEL is orthogonal to that
Sybil-resistance: it controls whether a ballot reveals the voter's persona, never whether one human
can stuff the box.

IdentityTier (the "various levels of anonymity"):
  * ANONYMOUS       — the ballot is signed by a FRESH EPHEMERAL key, unlinkable to any persona; only
                      the nullifier ties it to "one human" (and the personhood layer keeps that
                      nullifier unlinkable across polls). You learn the tally, never who chose what.
                      Eligibility (the nullifier belongs to an enrolled human) is proven in production
                      by an anonymous credential (realid/ps_credential); here that check is a seam.
  * PSEUDONYMOUS    — the ballot is signed by the voter's persona; the choice is visible under that nym.
  * VERIFIED_PERSON — like pseudonymous, but the nullifier is a PER-HUMAN tag (personhood-backed), so
                      one human = one response even across their pseudonyms. Accountable; real-ID hidden.

Not new crypto — HybridSig over domain-separated, length-prefixed bodies + nullifier dedup.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Sequence, Tuple

from ..crypto.primitives import H
from ..crypto.sign import HybridSigKeypair, HybridSigPublic, sign, verify
from .kinds import IdentityTier   # the canonical anonymity axis (shared with Spaces)

__all__ = ["IdentityTier", "Poll", "PollResponse", "PollResult", "create_poll", "verify_poll",
           "respond", "respond_anonymously", "verify_response", "tally"]

_POLL_DOMAIN = b"atlas/poll/v1"
_RESPONSE_DOMAIN = b"atlas/poll-response/v1"


def _lp(b: bytes) -> bytes:
    return len(b).to_bytes(4, "big") + b


# --------------------------------------------------------------------------- poll
@dataclass
class Poll:
    author: HybridSigPublic
    question: bytes
    options: Tuple[bytes, ...]
    tier: IdentityTier
    epoch: int
    sig: bytes = b""

    def _body(self) -> bytes:
        parts = [_POLL_DOMAIN, _lp(self.author.encode()), _lp(self.question),
                 len(self.options).to_bytes(4, "big")]
        parts.extend(_lp(o) for o in self.options)
        parts.append(int(self.tier).to_bytes(2, "big"))
        parts.append(self.epoch.to_bytes(8, "big"))
        return b"".join(parts)

    def poll_id(self) -> bytes:
        """A stable id derived from the poll's content (author + question + options + tier + epoch)."""
        return H(b"atlas/poll-id", self._body())


def create_poll(kp: HybridSigKeypair, *, question: bytes, options: Sequence[bytes],
                tier: IdentityTier, epoch: int) -> Poll:
    if len(options) < 2:
        raise ValueError("a poll needs >= 2 options")
    p = Poll(author=kp.public, question=question, options=tuple(options), tier=tier, epoch=epoch)
    p.sig = sign(kp, p._body())
    return p


def verify_poll(p: Poll) -> bool:
    return len(p.options) >= 2 and verify(p.author, p._body(), p.sig)


# --------------------------------------------------------------------------- response (ballot)
@dataclass
class PollResponse:
    poll_id: bytes
    choice: int                    # index into poll.options
    nullifier: bytes               # one-human-one-response dedup key (from the personhood layer)
    ballot_key: HybridSigPublic    # voter persona (pseudonymous/verified) OR ephemeral (anonymous)
    epoch: int
    sig: bytes = b""

    def _body(self) -> bytes:
        return b"".join([_RESPONSE_DOMAIN, _lp(self.poll_id), self.choice.to_bytes(4, "big"),
                         _lp(self.nullifier), _lp(self.ballot_key.encode()),
                         self.epoch.to_bytes(8, "big")])


def _make_response(poll: Poll, choice: int, nullifier: bytes, ballot_key: HybridSigPublic,
                   epoch: int) -> PollResponse:
    if not (0 <= choice < len(poll.options)):
        raise ValueError("choice out of range")
    return PollResponse(poll_id=poll.poll_id(), choice=choice, nullifier=nullifier,
                        ballot_key=ballot_key, epoch=epoch)


def respond(voter_kp: HybridSigKeypair, poll: Poll, *, choice: int, nullifier: bytes,
            epoch: int) -> PollResponse:
    """A PSEUDONYMOUS / VERIFIED_PERSON ballot — signed by the voter's persona (the choice is visible
    under that nym). One-human-one-response via `nullifier`; for VERIFIED_PERSON the nullifier is a
    per-human tag, so it holds even across the voter's pseudonyms."""
    r = _make_response(poll, choice, nullifier, voter_kp.public, epoch)
    r.sig = sign(voter_kp, r._body())
    return r


def respond_anonymously(poll: Poll, *, choice: int, nullifier: bytes, epoch: int,
                        ephemeral_kp: HybridSigKeypair) -> PollResponse:
    """An ANONYMOUS ballot — signed by a FRESH EPHEMERAL key that is unlinkable to the voter's persona.
    The choice can never be tied back to a person; only the `nullifier` enforces one-human-one-response
    (unlinkable across polls, from the personhood layer). Eligibility — that this ballot is backed by an
    enrolled human — is proven by an anonymous credential (realid/ps_credential) in production."""
    r = _make_response(poll, choice, nullifier, ephemeral_kp.public, epoch)
    r.sig = sign(ephemeral_kp, r._body())
    return r


def verify_response(poll: Poll, r: PollResponse) -> bool:
    return (r.poll_id == poll.poll_id() and 0 <= r.choice < len(poll.options)
            and verify(r.ballot_key, r._body(), r.sig))


# --------------------------------------------------------------------------- tally
@dataclass(frozen=True)
class PollResult:
    poll_id: bytes
    counts: Tuple[int, ...]   # votes per option, aligned to poll.options
    total: int                # distinct humans who responded

    def winner(self) -> int:
        """Index of the leading option (first, on a tie)."""
        return max(range(len(self.counts)), key=lambda i: self.counts[i]) if self.counts else -1


def tally(poll: Poll, responses: Sequence[PollResponse]) -> PollResult:
    """One-human-one-response: dedupe by `nullifier` (LAST valid response wins, so a voter can change
    their choice — it flips, never stacks). Only valid responses FOR THIS poll count."""
    latest: Dict[bytes, PollResponse] = {}
    for r in responses:
        if verify_response(poll, r):
            latest[r.nullifier] = r
    counts = [0] * len(poll.options)
    for r in latest.values():
        counts[r.choice] += 1
    return PollResult(poll_id=poll.poll_id(), counts=tuple(counts), total=len(latest))
